File: apps/native-host/internal/protocol/protocol.go
package protocol

import "encoding/json"

const Version = 1

type Request struct {
	ProtocolVersion int             `json:"protocolVersion"`
	RequestID       string          `json:"requestId"`
	Method          string          `json:"method"`
	Params          json.RawMessage `json:"params"`
}

type Error struct {
	Code    string         `json:"code"`
	Message string         `json:"message"`
	Details map[string]any `json:"details,omitempty"`
}

type Response struct {
	ProtocolVersion int    `json:"protocolVersion"`
	RequestID       string `json:"requestId"`
	OK              bool   `json:"ok"`
	Result          any    `json:"result,omitempty"`
	Error           *Error `json:"error,omitempty"`
}

type LocalToolStatus struct {
	Installed         bool   `json:"installed"`
	Version           string `json:"version,omitempty"`
	Ready             bool   `json:"ready"`
	Authenticated     bool   `json:"authenticated"`
	Configured        bool   `json:"configured"`
	TargetType        string `json:"targetType,omitempty"`
	TargetDisplayName string `json:"targetDisplayName,omitempty"`
	Message           string `json:"message"`
}

type LocalOpenParams struct {
	Action string `json:"action"`
}

type LocalOpenResult struct {
	Started bool   `json:"started"`
	Action  string `json:"action"`
	Message string `json:"message"`
}

type HealthResult struct {
	HostVersion     string          `json:"hostVersion"`
	ProtocolVersion int             `json:"protocolVersion"`
	Platform        string          `json:"platform"`
	Capabilities    []string        `json:"capabilities"`
	LocalTool       LocalToolStatus `json:"localTool"`
}

type PageSnapshot struct {
	URL               string `json:"url,omitempty"`
	Title             string `json:"title,omitempty"`
	VisibleTextBytes  int    `json:"visibleTextBytes,omitempty"`
	VisibleBlockCount int    `json:"visibleBlockCount,omitempty"`
	VisibleImageCount int    `json:"visibleImageCount,omitempty"`
	ExportRequired    bool   `json:"exportRequired"`
}

type MigrationExportParams struct {
	FileName      string        `json:"fileName"`
	ContentBase64 string        `json:"contentBase64"`
	Title         string        `json:"title,omitempty"`
	SubpageMode   string        `json:"subpageMode,omitempty"`
	CreateNew     bool          `json:"createNew,omitempty"`
	SourcePage    *PageSnapshot `json:"sourcePage,omitempty"`
}

type MigrationExportInspection struct {
	FileName       string                   `json:"fileName"`
	Title          string                   `json:"title"`
	ExportedAt     string                   `json:"exportedAt"`
	PageCount      int                      `json:"pageCount"`
	Bytes          int                      `json:"bytes"`
	PreviousExport *MigrationPreviousExport `json:"previousExport,omitempty"`
}

type MigrationPreviousExport struct {
	Status      string `json:"status"`
	DocumentURL string `json:"documentUrl"`
	UpdatedAt   string `json:"updatedAt"`
	Count       int    `json:"count"`
	SubpageMode string `json:"subpageMode"`
}

type MigrationResult struct {
	TaskID                    string `json:"taskId"`
	RemoteTaskID              string `json:"remoteTaskId,omitempty"`
	DocumentURL               string `json:"documentUrl"`
	Reused                    bool   `json:"reused"`
	ExpectedImageCount        int    `json:"expectedImageCount"`
	ReadbackImageCount        int    `json:"readbackImageCount"`
	NativeTodoCount           int    `json:"nativeTodoCount"`
	NativeCodeBlockCount      int    `json:"nativeCodeBlockCount"`
	NativeLayoutCount         int    `json:"nativeLayoutCount"`
	NativeSubpageTocItemCount int    `json:"nativeSubpageTocItemCount"`
	SubpageMode               string `json:"subpageMode,omitempty"`
	RecursivePageCount        int    `json:"recursivePageCount,omitempty"`
	RecursiveFolderCount      int    `json:"recursiveFolderCount,omitempty"`
	RecursiveLinkCount        int    `json:"recursiveLinkCount,omitempty"`
	CleanupVerified           bool   `json:"cleanupVerified"`
	SourcePageCaptured        bool   `json:"sourcePageCaptured"`
}

type MigrationTaskParams struct {
	TaskID string `json:"taskId"`
}

type MigrationTaskProgress struct {
	Current int    `json:"current"`
	Total   int    `json:"total"`
	Percent int    `json:"percent"`
	Stage   string `json:"stage"`
	Message string `json:"message"`
}

type MigrationTaskSnapshot struct {
	TaskID         string                `json:"taskId"`
	Status         string                `json:"status"`
	Progress       MigrationTaskProgress `json:"progress"`
	CanCancel      bool                  `json:"canCancel"`
	RecoveryAction string                `json:"recoveryAction,omitempty"`
	UpdatedAt      string                `json:"updatedAt"`
	Result         *MigrationResult      `json:"result,omitempty"`
	Error          *Error                `json:"error,omitempty"`
}

func Failure(requestID, code, message string) Response {
	return Response{
		ProtocolVersion: Version,
		RequestID:       requestID,
		OK:              false,
		Error: &Error{
			Code:    code,
			Message: message,
		},
	}
}

func FailureWithDetails(requestID, code, message string, details map[string]any) Response {
	response := Failure(requestID, code, message)
	response.Error.Details = details
	return response
}

File: apps/edge-extension/src/protocol.ts
export const NATIVE_HOST_NAME = "com.leonz03.notion2dingding";
export const PROTOCOL_VERSION = 1 as const;

export interface NativeRequest<TParams = unknown> {
  protocolVersion: typeof PROTOCOL_VERSION;
  requestId: string;
  method: string;
  params: TParams;
}

export interface NativeError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export interface NativeResponse<TResult = unknown> {
  protocolVersion: typeof PROTOCOL_VERSION;
  requestId: string;
  ok: boolean;
  result?: TResult;
  error?: NativeError;
}

export interface HealthResult {
  hostVersion: string;
  protocolVersion: typeof PROTOCOL_VERSION;
  platform: string;
  capabilities: string[];
  localTool: {
    installed: boolean;
    version?: string;
    ready: boolean;
    authenticated: boolean;
    configured: boolean;
    targetType?: "folder" | "workspace";
    targetDisplayName?: string;
    message: string;
  };
}

export interface LocalOpenParams {
  action: "login" | "target";
}

export interface LocalOpenResult {
  started: boolean;
  action: "login" | "target";
  message: string;
}

export interface PageSnapshot {
  url: string;
  title: string;
  visibleTextBytes: number;
  visibleBlockCount: number;
  visibleImageCount: number;
  exportRequired: true;
}

export interface MigrationExportParams {
  fileName: string;
  contentBase64: string;
  title?: string;
  subpageMode?: "inline" | "tree";
  createNew?: boolean;
  sourcePage?: PageSnapshot;
}

export interface MigrationExportInspection {
  fileName: string;
  title: string;
  exportedAt: string;
  pageCount: number;
  bytes: number;
  previousExport?: MigrationPreviousExport;
}

export interface MigrationPreviousExport {
  status: "success" | "unknown";
  documentUrl: string;
  updatedAt: string;
  count: number;
  subpageMode: "inline" | "tree";
}

export interface MigrationResult {
  taskId: string;
  remoteTaskId?: string;
  documentUrl: string;
  reused: boolean;
  expectedImageCount: number;
  readbackImageCount: number;
  nativeTodoCount: number;
  nativeCodeBlockCount?: number;
  nativeLayoutCount?: number;
  nativeSubpageTocItemCount?: number;
  subpageMode?: "inline" | "tree";
  recursivePageCount?: number;
  recursiveFolderCount?: number;
  recursiveLinkCount?: number;
  cleanupVerified: boolean;
  sourcePageCaptured: boolean;
}

export interface MigrationTaskParams {
  taskId: string;
}

export type MigrationTaskStatus =
  | "queued"
  | "running"
  | "cancel_requested"
  | "succeeded"
  | "failed"
  | "unknown"
  | "cancelled";

export interface MigrationTaskProgress {
  current: number;
  total: number;
  percent: number;
  stage: string;
  message: string;
}

export interface MigrationTaskSnapshot {
  taskId: string;
  status: MigrationTaskStatus;
  progress: MigrationTaskProgress;
  canCancel: boolean;
  recoveryAction?: "login" | "target" | "select_export" | "retry" | "recover";
  updatedAt: string;
  result?: MigrationResult;
  error?: NativeError;
}

export function createRequest<TParams>(
  method: string,
  params: TParams,
): NativeRequest<TParams> {
  return {
    protocolVersion: PROTOCOL_VERSION,
    requestId: crypto.randomUUID(),
    method,
    params,
  };
}

export function isNativeResponse(value: unknown): value is NativeResponse {
  if (typeof value !== "object" || value === null) {
    return false;
  }

  const candidate = value as Partial<NativeResponse>;
  const envelopeValid =
    candidate.protocolVersion === PROTOCOL_VERSION &&
    typeof candidate.requestId === "string" &&
    typeof candidate.ok === "boolean";
  if (!envelopeValid) {
    return false;
  }
  if (candidate.ok) {
    return candidate.error === undefined;
  }
  return (
    typeof candidate.error?.code === "string" &&
    typeof candidate.error.message === "string"
  );
}
